### ISSUE TITLE
Lint warnings for "unused #[must_use]" should be emitted even if typeck fails.

### ISSUE THREAD

**Issue by eddyb** [rust-lang/rust#51825]
In a situation like this (feel free to suggest less contrived examples):
```rust
#[must_use]
struct Foo<T> {
    text: String,
    children: Vec<T>,
}

impl<T> Foo<T> {
    fn combine(mut self, other: Self) -> Self {
        self.text.push_str(&other.text);
        self.children.extend(other.children);
        self
    }
}

fn text<T>(text: impl Into<String>) -> Foo<T> {
    Foo {
        text: text.into(),
        children: vec![],
    }
}

fn bar<T>(mut foo: Foo<T>) -> Foo<T> {
    if foo.text.is_empty() {
        // Incomplete/incorrect code, #[must_use] should warn.
        // However, we already have a type inference error.
        text("bar");
        // Correct code:
        foo = text("bar").combine(foo);
    }
    foo
}
```
Because of the type-checking error caused by an uninferrable `T`, the `#[must_use]` lint never gets to run. But the fact that the return value isn't used is, in this case, responsible for the error.

If the `#[must_use]` lint is moved to typeck, it could run even on incomplete inference. Alternatively, the lint could stay independent, but somehow always run, even if typeck errored. The `test("bar")` expression would need to have the type `Foo<_>` with the `_` being a `TyError`.

cc @nikomatsakis @Manishearth
